### PR TITLE
feat: add sm2 encrypted test case from GM/T0003.5-2012 to test suite. And fix some place.

### PR DESCRIPTION
**Pull request by ymj-uos**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — COMMENTED** [review]
This would be better added to `test/recipes/30-test_evp_data/evppkey_sm2.txt`

**Comment by ymj-uos** [comment]
in the evppkey_sm2.txt, it does not set random data, and it is not better to do the encryption and decryption.

File: test/sm2_internal_test.c [hunk 1]
```diff
@@ -184,7 +184,7 @@ static int test_sm2_crypt(const EC_GROUP *group,
         goto done;
 
     if (!TEST_true(ossl_sm2_plaintext_size(key, digest, ctext_len, &ptext_len))
-            || !TEST_int_eq(ptext_len, msg_len))
+            || !TEST_int_ge(ptext_len, msg_len))
```
Comment (paulidale) [inline test/sm2_internal_test.c:187]: Why would the plaintext length ever be larger than the message length?

Comment (ymj-uos) [reply test/sm2_internal_test.c:187]: int ossl_sm2_plaintext_size(const EC_KEY *key, const EVP_MD *digest,
                            size_t msg_len, size_t *pt_size)
{
    const size_t field_size = ec_field_size(EC_KEY_get0_group(key));
    const int md_size = EVP_MD_get_size(digest);
    size_t overhead;

    if (md_size < 0) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_DIGEST);
        return 0;
    }
    if (field_size == 0) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_FIELD);
        return 0;
    }

    **overhead = 10 + 2 * field_size + (size_t)md_size;**
    if (msg_len <= overhead) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_ENCODING);
        return 0;
    }

    *pt_size = msg_len - overhead;
    return 1;
}

Comment (ymj-uos) [reply test/sm2_internal_test.c:187]: int ossl_sm2_plaintext_size(const EC_KEY *key, const EVP_MD *digest,
                            size_t msg_len, size_t *pt_size)
{
    const size_t field_size = ec_field_size(EC_KEY_get0_group(key));
    const int md_size = EVP_MD_get_size(digest);
    size_t overhead;

    if (md_size < 0) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_DIGEST);
        return 0;
    }
    if (field_size == 0) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_FIELD);
        return 0;
    }

    overhead = 10 + 2 * field_size + (size_t)md_size;
    if (msg_len <= overhead) {
        ERR_raise(ERR_LIB_SM2, SM2_R_INVALID_ENCODING);
        return 0;
    }

    *pt_size = msg_len - overhead;
    return 1;
}

Comment (ymj-uos) [reply test/sm2_internal_test.c:187]: overhead = 10 + 2 * field_size + (size_t)md_size;
10 is not Fixed value, sometimes is larger then 10.
it depends on the asn.1 structure.
in this case, the asn.1 structure len is equal to 11 (not 10).

**Comment by ymj-uos** [comment]
image

**Comment by kaduk** [comment]
close/reopen to re-trigger the CLA check

**Comment by romen** [comment]
It seems the CLA check is still failing. @mattcaswell can you check this?

**Comment by ymj-uos** [comment]
image

**Comment by ymj-uos** [comment]
How to change the email to the **_yangmingjun@uniontech.com_**!

**Comment by mspncp** [comment]
> 
> 
> How to change the email to the **_[yangmingjun@uniontech.com](mailto:yangmingjun@uniontech.com)_**!

You can edit your personal e-mail address(es) at https://github.com/settings/emails .

**Comment by ymj-uos** [comment]
> > How to change the email to the **_[yangmingjun@uniontech.com](mailto:yangmingjun@uniontech.com)_**!
> 
> You can edit your personal e-mail address(es) at https://github.com/settings/emails .



> > How to change the email to the **_[yangmingjun@uniontech.com](mailto:yangmingjun@uniontech.com)_**!
> 
> You can edit your personal e-mail address(es) at https://github.com/settings/emails .

image

**Comment by mspncp** [comment]
If you checked the option to keep your email addresses private, you need to uncheck it again:

grafik

**Comment by mspncp** [comment]
Your settings look good. Did set them just now or where they already set like this?

**Comment by ymj-uos** [comment]
uncheck it! and what next... I have tried some ways, but not correct.

**Comment by mspncp** [comment]
Changing the author email of an existing commit is not supported by the GitHub web interface. It needs to be done locally using the git commandline tool.  Your pull request branch needs to be rebased anyway, because you have a merge commit from master,  975845d5c5f0881863be2493a7b91358d2ab3fad, which we don't allow.

If you wan't, I can do the rebase and force-push it to your branch. All you need to do then is to reset your local branch. Please let me know if you wan't me to do it.

**Comment by mspncp** [comment]
I found your email address twice in the git log. If you want me to reset the author address of your commits, which of the two versions do you want me to use?
```
Author: Mingjun.Yang <yangmingjun@uniontech.com>
Author: 杨明君 <yangmingjun@uniontech.com>
```

**Comment by ymj-uos** [comment]
> Changing the author email of an existing commit is not supported by the GitHub web interface. It needs to be done locally using the git commandline tool. Your pull request branch needs to be rebased anyway, because you have a merge commit from master, [975845d](https://github.com/openssl/openssl/commit/975845d5c5f0881863be2493a7b91358d2ab3fad), which we don't allow.
> 
> If you wan't, I can do the rebase and force-push it to your remote branch. All you need to do then is to reset your local branch. Please let me know if you wan't me to do it.

thanks! you can rebase my branch.

**Comment by ymj-uos** [comment]
>  Mingjun.Yang <yangmingjun@uniontech.com>

 Mingjun.Yang <yangmingjun@uniontech.com> 

this one.

**Comment by mspncp** [comment]
There is a merge conflict in `test_sm2_crypt()`. Variant A is on 'master', variant B on your branch. What is the correct merged version?

```
<<<<<<< variant A
    if (!TEST_true(ossl_sm2_plaintext_size(ctext, ctext_len, &ptext_len))
            || !TEST_int_eq(ptext_len, msg_len))
>>>>>>> variant B
    if (!TEST_true(ossl_sm2_plaintext_size(key, digest, ctext_len, &ptext_len))
            || !TEST_int_ge(ptext_len, msg_len))
####### Ancestor
    if (!TEST_true(ossl_sm2_plaintext_size(key, digest, ctext_len, &ptext_len))
            || !TEST_int_eq(ptext_len, msg_len))
======= end
```

**Comment by mspncp** [comment]
I guess it's

```
    if (!TEST_true(ossl_sm2_plaintext_size(ctext, ctext_len, &ptext_len))
            || !TEST_int_ge(ptext_len, msg_len))
```
but I'm not sure.

**Comment by ymj-uos** [comment]
> if (!TEST_true(ossl_sm2_plaintext_size(ctext, ctext_len, &ptext_len))
>             || !TEST_int_eq(ptext_len, msg_len))

please keep the origianl version!
if (!TEST_true(ossl_sm2_plaintext_size(ctext, ctext_len, &ptext_len))
            || !TEST_int_eq(ptext_len, msg_len))

During uploading my codes, the api is changed! so I fixed it.

**Comment by mspncp** [comment]
I rebased your branch and squashed the fixups into a single commit. Please reset your local branch and check the result.

**Comment by mspncp** [comment]
You just overwrote my changes with your force-push. I am going to force-push again...

**Comment by mspncp** [comment]
Sorry, something changed, I don't seem to have the permission to push to your branch anymore.

**Comment by mspncp** [comment]
Maybe it's better if you close this pr and open a new one?  You find my rebased and squashed commit at bc37353a3649bd42e2c7b80a760fe55023812328.

IMHO It's better if you create a topic branch for your pull request and don't reuse the 'master' branch of your fork.

**Comment by ymj-uos** [comment]
@mspncp I open a new one.

**Comment by romen** [comment]
Closed in favour of #16511

**Pull request closed without merging**